gpkit/model.py

```diff
@@ -1,250 +1,424 @@
+            for p in posynomials:
+                for k in p.varlocs:
+                    if "model" not in k.descr:
+                        newk = VarKey(k, model=name)
+                        p.varlocs[newk] = p.varlocs.pop(k)
+                for exp in p.exps:
+                    for k in exp:
+                        if "model" not in k.descr:
+                            newk = VarKey(k, model=name)
+                            exp[newk] = exp.pop(k)
+
+    @property
+    def variables(self):
+        variables = {}
+        for p in self.posynomials:
+            variables.update({vk: Variable(**vk.descr) for vk in p.varlocs})
+        return variables
+
+    @property
+    def posynomials(self):
+        constraints = tuple(flatten(self.constraints, Constraint))
+        # TODO: parse constraints during flattening, calling Posyarray on
+        #       anything that holds only posys and then saving that list.
+        #       This will allow prettier constraint printing.
+        posynomials = [self.cost]
+        for constraint in constraints:
+            if isinstance(constraint, MonoEQConstraint):
+                posynomials += [constraint.leq, constraint.geq]
+            else:
+                posynomials.append(constraint)
+        return posynomials
+
+    def __add__(self, other):
+        if isinstance(other, Model):
+            substitutions = dict(self.substitutions)
+            substitutions.update(other.substitutions)
+            return Model(self.cost*other.cost,
+                         self.constraints + other.constraints,
+                         substitutions)
+        else:
+            return NotImplemented
+
+    def __mul__(self, other):
+        if isinstance(other, Model):
+            # TODO: combine shared variables
+            substitutions = dict(self.substitutions)
+            substitutions.update(other.substitutions)
+            return Model(self.cost*other.cost,
+                         self.constraints + other.constraints,
+                         substitutions)
+        else:
+            return NotImplemented
 
-    def __repr__(self):
-        return "\n".join(["gpkit.Model with",
-                          "  Equations"] +
-                         ["     %s <= 1" % p._string()
-                          for p in self.posynomials])
-
-    def add_constraints(self, constraints):
-        if isinstance(constraints, Constraint):
-            constraints = [constraints]
-        self.constraints += tuple(constraints)
-        self._gen_unsubbed_vars()
-
-    def rm_constraints(self, constraints):
-        if isinstance(constraints, Constraint):
-            constraints = [constraints]
-        for p in constraints:
-            self.constraints.remove(p)
-        self._gen_unsubbed_vars()
-
-    def _gen_unsubbed_vars(self, exps=None, cs=None):
+    # TODO: add get_item
+
+    def solve(self, solver=None, verbosity=2, skipfailures=True, *args, **kwargs):
```

Comment:
Probably need a docstring here -- how does this relate to `GeometricProgram.solve` from a user's point of view?